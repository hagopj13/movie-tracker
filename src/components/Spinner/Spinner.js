import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    margin: theme.spacing(3, 0),
  },
}));

const Spinner = () => {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <CircularProgress color="primary" />
    </div>
  );
};

export default Spinner;
