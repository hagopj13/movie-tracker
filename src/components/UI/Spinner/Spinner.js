import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    margin: theme.spacing(3, 0),
  },
}));

export default function CircularIndeterminate() {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <CircularProgress color="primary" />
    </div>
  );
}
