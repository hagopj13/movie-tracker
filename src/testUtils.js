import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { createMount } from '@material-ui/core/test-utils';

import { initialState } from 'store/rootReducer';

const mount = createMount();
const mockStore = configureMockStore()(initialState);

export const mountWithStore = (children, store = mockStore) =>
  mount(<Provider store={store}>{children}</Provider>);
