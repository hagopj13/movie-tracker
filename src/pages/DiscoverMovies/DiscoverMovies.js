// @flow
import React, { useLayoutEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import Spinner from 'components/Spinner/Spinner';
import Attribution from 'components/Attribution/Attribution';
import discoverActions from 'store/discover/discover.actions';
import loadingSelectors from 'store/api/loading/loading.selectors';
import ConfigActionTypes from 'store/config/config.types';
import configActions from 'store/config/config.actions';

import DiscoverMoviesList from './List/DiscoverMoviesList';
import DiscoverMoviesFilters from './Filters/DiscoverMoviesFilters';

type Props = {
  isLoadingConfig: boolean,
  onFetchAllGenres: () => void,
  onFetchMovies: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    flexDirection: 'column',
  },
  attributionContainer: {
    display: 'flex',
    justifyContent: 'center',
    padding: theme.spacing(5, 5, 0),
  },
  moviesContainer: {
    display: 'flex',
    padding: theme.spacing(5, 5, 2),
    [theme.breakpoints.down('xs')]: {
      flexDirection: 'column',
      alignItems: 'center',
    },
  },
  movieFiltersContainer: {
    maxWidth: 270,
    marginRight: theme.spacing(3),
    [theme.breakpoints.down('xs')]: {
      maxWidth: 330,
      marginRight: 0,
      marginBottom: theme.spacing(3),
    },
  },
  movieListContainer: {
    flexGrow: 1,
  },
}));

export const DiscoverMovies = (props: Props) => {
  const { isLoadingConfig, onFetchAllGenres, onFetchMovies } = props;

  const classes = useStyles();

  useLayoutEffect(() => {
    onFetchAllGenres();
    onFetchMovies();
  }, [onFetchAllGenres, onFetchMovies]);

  if (isLoadingConfig) {
    return <Spinner />;
  }

  return (
    <div className={classes.root}>
      <Container className={classes.attributionContainer}>
        <Attribution />
      </Container>
      <Container className={classes.moviesContainer}>
        <div className={classes.movieFiltersContainer}>
          <DiscoverMoviesFilters />
        </div>
        <div className={classes.movieListContainer}>
          <DiscoverMoviesList />
        </div>
      </Container>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  isLoadingConfig: loadingSelectors.createIsLoadingSelector([
    ConfigActionTypes.FETCH_IMAGES_CONFIG,
    ConfigActionTypes.FETCH_ALL_GENRES,
  ]),
});

const mapDispatchToProps = {
  onFetchAllGenres: configActions.fetchAllGenres,
  onFetchMovies: discoverActions.fetchMovies,
};

export default connect(mapStateToProps, mapDispatchToProps)(DiscoverMovies);
