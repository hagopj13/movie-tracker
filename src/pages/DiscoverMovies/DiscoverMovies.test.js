import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { mountWithStore } from 'testUtils';
import Spinner from 'components/Spinner/Spinner';

import { DiscoverMovies } from './DiscoverMovies';

describe('DiscoverMovies page', () => {
  let shallow;
  let wrapper;

  const mockFetchAllGenres = jest.fn();
  const mockFetchMovies = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the DiscoverMovies page correctly', () => {
    const mockProps = {
      isLoadingConfig: false,
      onFetchAllGenres: mockFetchAllGenres,
      onFetchMovies: mockFetchMovies,
    };
    wrapper = shallow(<DiscoverMovies {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Spinner when config is loading', () => {
    const mockProps = {
      isLoadingConfig: true,
      onFetchAllGenres: mockFetchAllGenres,
      onFetchMovies: mockFetchMovies,
    };
    wrapper = shallow(<DiscoverMovies {...mockProps} />);
    expect(wrapper.exists(Spinner)).toBe(true);
  });

  it('should fetch all genres and movies after mounting', () => {
    const mockProps = {
      isLoadingConfig: false,
      onFetchAllGenres: mockFetchAllGenres,
      onFetchMovies: mockFetchMovies,
    };
    wrapper = mountWithStore(<DiscoverMovies {...mockProps} />);
    expect(mockFetchAllGenres).toHaveBeenCalledTimes(1);
    expect(mockFetchMovies).toHaveBeenCalledTimes(1);
  });
});
