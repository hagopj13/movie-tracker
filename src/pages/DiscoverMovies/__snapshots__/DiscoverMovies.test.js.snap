// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscoverMovies page should render the DiscoverMovies page correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-attributionContainer-2"
  >
    <Attribution />
  </WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-moviesContainer-3"
  >
    <div
      className="makeStyles-movieFiltersContainer-4"
    >
      <Connect(DiscoverMoviesFilters) />
    </div>
    <div
      className="makeStyles-movieListContainer-5"
    >
      <Connect(DiscoverMoviesList) />
    </div>
  </WithStyles(ForwardRef(Container))>
</div>
`;
